src/evaluator.ts

```diff
@@ -573,78 +584,114 @@ class Evaluator {
-          varNames.add(varName);
-          varWithSpecialColsRemoved[varName] = testCase.vars[varName];
+    for (const mainTestCase of tests) {
+      const allTestCases = mainTestCase.thread ? [...mainTestCase.thread, mainTestCase] : [mainTestCase]
+      for (const testCase of allTestCases) {
+        if (testCase.vars) {
+          const varWithSpecialColsRemoved: Record<string, string | string[] | object> = {};
+          for (const varName of Object.keys(testCase.vars)) {
+            varNames.add(varName);
+            varWithSpecialColsRemoved[varName] = testCase.vars[varName];
+          }
+          varsWithSpecialColsRemoved.push(varWithSpecialColsRemoved);
         }
-        varsWithSpecialColsRemoved.push(varWithSpecialColsRemoved);
       }
     }
 
     // Set up eval cases
-    let runEvalOptions: RunEvalOptions[] = [];
+    let runEvalOptions: RunEvalOptions[][] = [];
     let rowIndex = 0;
     for (let index = 0; index < tests.length; index++) {
-      const testCase = tests[index];
-      invariant(
-        Array.isArray(testSuite.defaultTest?.assert || []),
-        `defaultTest.assert is not an array in test case #${index + 1}`,
-      );
-      invariant(
-        Array.isArray(testCase.assert || []),
-        `testCase.assert is not an array in test case #${index + 1}`,
-      );
-      // Handle default properties
-      testCase.vars = { ...testSuite.defaultTest?.vars, ...testCase.vars };
-      testCase.assert = [...(testSuite.defaultTest?.assert || []), ...(testCase.assert || [])];
-      testCase.threshold = testCase.threshold ?? testSuite.defaultTest?.threshold;
-      testCase.options = { ...testSuite.defaultTest?.options, ...testCase.options };
-
-      const prependToPrompt =
-        testCase.options?.prefix || testSuite.defaultTest?.options?.prefix || '';
-      const appendToPrompt =
-        testCase.options?.suffix || testSuite.defaultTest?.options?.suffix || '';
-
-      // Finalize test case eval
-      const varCombinations =
-        process.env.PROMPTFOO_DISABLE_VAR_EXPANSION || testCase.options.disableVarExpansion
-          ? [testCase.vars]
-          : generateVarCombinations(testCase.vars || {});
-
-      const numRepeat = this.options.repeat || 1;
-      for (let repeatIndex = 0; repeatIndex < numRepeat; repeatIndex++) {
-        for (const vars of varCombinations) {
-          let colIndex = 0;
-          for (const prompt of testSuite.prompts) {
-            for (const provider of testSuite.providers) {
-              if (testSuite.providerPromptMap) {
-                const allowedPrompts = testSuite.providerPromptMap[provider.id()];
-                if (allowedPrompts && !allowedPrompts.includes(prompt.display)) {
-                  // This prompt should not be used with this provider.
-                  continue;
+      const mainTestCase = tests[index];
+      const allTestCases = mainTestCase.thread || [mainTestCase];
+      const startRow = rowIndex
```

Comment:
If we're in a threaded context, we will not add tests column-by-column. The tests have to be added row-by-row. This means that if we have the following config:

```
providers: [gpt-3, gpt-4]

tests:
  - thread:
      - assertions: ...
      - assertions: ...
 - assertions: ...
```

We will fill the table in the following order:
```
+-----------+--------+
| Outputs   |        |
+-----------+--------+
| 1         | 3      |
+-----------+--------+
| 2         | 4      |
+-----------+--------+
| 5         | 6      |
+-----------+--------+
```